File: go/builtins/funcEndsWith.go
package builtins

import (
	"strings"

	"github.com/jplorg/jpl/go/jpl"
	"github.com/jplorg/jpl/go/library"
)

var funcEndsWith = enclose(func(runtime jpl.JPLRuntime, signal jpl.JPLRuntimeSignal, next jpl.JPLPiper, input any, args ...any) ([]any, error) {
	value, err := library.UnwrapValue(input)
	if err != nil {
		return nil, err
	}
	t, err := library.Type(value)
	if err != nil {
		return nil, err
	}

	var token any
	if len(args) > 0 {
		token, err = library.UnwrapValue(args[0])
		if err != nil {
			return nil, err
		}
	}
	tt, err := library.Type(token)
	if err != nil {
		return nil, err
	}

	switch t {
	case jpl.JPLT_STRING:
		if tt == jpl.JPLT_STRING {
			return next.Pipe(strings.HasSuffix(value.(string), token.(string)))
		}

	default:
	}

	return nil, library.ThrowAny(library.NewTypeError("%s (%*<100v) cannot end with %s (%*<100v)", string(t), value, string(tt), token))
})

File: js/lib/builtins/funcEndsWith.ts
import type JPLRuntimeSignal from '@/library/runtimeSignal';
import type JPLRuntime from '@/runtime';
import { JPLTypeError } from '../library';

export default function builtin(
  runtime: JPLRuntime,
  _signal: JPLRuntimeSignal,
  next: (output: unknown) => Promise<unknown[]> | unknown[],
  input: unknown,
  arg0: unknown,
): Promise<unknown[]> | unknown[] {
  const value = runtime.unwrapValue(input);
  const t = runtime.type(value);

  const token = runtime.unwrapValue(arg0 ?? null);
  const tt = runtime.type(token);

  switch (t) {
    case 'string':
      if (tt === 'string')
        return next((value as string).endsWith(token as string));
      break;

    default:
  }

  throw new JPLTypeError(
    '%s (%*<100v) cannot end with %s (%*<100v)',
    t,
    value,
    tt,
    token,
  );
}
